File: lib/src/main/java/com/yandex/authsdk/internal/JwtRequest.java
package com.yandex.authsdk.internal;

import androidx.annotation.NonNull;

import com.yandex.authsdk.YandexAuthException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class JwtRequest {

    public static final String JWT_REQUEST_URL_FORMAT = "https://login.yandex.ru/info?format=jwt&oauth_token=%s";

    public static final int RESPONSE_CODE_UNAUTHORIZED = 401;

    @NonNull
    private final String token;

    public JwtRequest(@NonNull final String token) {
        this.token = token;
    }

    @NonNull
    public String get() throws IOException, YandexAuthException {
        final URL url = new URL(String.format(JWT_REQUEST_URL_FORMAT, token));
        final HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        if (connection.getResponseCode() == RESPONSE_CODE_UNAUTHORIZED) {
            throw new YandexAuthException(YandexAuthException.JWT_AUTHORIZATION_ERROR);
        }
        try {
            return readToString(connection.getInputStream());
        } finally {
            connection.disconnect();
        }
    }

    @NonNull
    private static String readToString(@NonNull final InputStream is) throws IOException {
        final BufferedReader r = new BufferedReader(new InputStreamReader(is));
        try {
            final StringBuilder total = new StringBuilder();
            String line;
            while ((line = r.readLine()) != null) {
                total.append(line).append('\n');
            }
            return total.toString();
        } finally {
            r.close();
        }
    }
}

File: lib/src/main/java/com/yandex/authsdk/internal/JwtRequest.kt
package com.yandex.authsdk.internal

import com.yandex.authsdk.YandexAuthException
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

internal class JwtRequest(private val token: String) {

    @Throws(IOException::class, YandexAuthException::class)
    fun get(): String {
        val url = URL(String.format(JWT_REQUEST_URL_FORMAT, token))
        val connection = url.openConnection() as HttpURLConnection
        if (connection.responseCode == RESPONSE_CODE_UNAUTHORIZED) {
            throw YandexAuthException(YandexAuthException.JWT_AUTHORIZATION_ERROR)
        }
        return try {
            readToString(connection.inputStream)
        } finally {
            connection.disconnect()
        }
    }

    companion object {

        const val JWT_REQUEST_URL_FORMAT = "https://login.yandex.ru/info?format=jwt&oauth_token=%s"

        const val RESPONSE_CODE_UNAUTHORIZED = 401

        @Throws(IOException::class)
        private fun readToString(stream: InputStream): String {
            val reader = BufferedReader(InputStreamReader(stream))
            return reader.use { r ->
                val total = StringBuilder()
                var line: String?
                while (r.readLine().also { line = it } != null) {
                    total.append(line).append('\n')
                }
                total.toString()
            }
        }
    }
}
